### PR TITLE
Document how cases get into a hosted dataset

### PR DESCRIPTION
**Pull request by summerscope**
Follow-up to the concept gaps identified while revisiting #1876. That PR targeted `docs/evaluate/datasets/index.md`, which no longer exists after the evals docs restructure; this reworks the surviving concepts into the current page layout and corrects claims that no longer match the product.

## What was wrong

**Schemas read as advisory.** The docs said schemas "describe the data you want teammates to enter". They are enforced: every case is validated against the dataset's JSON schemas on add, update, and import, through the UI and SDK alike, and a non-matching request is rejected in full. Nothing warned a user that defining a schema can start rejecting writes, nor that existing cases are not rechecked when a schema changes, so a dataset can hold cases that would no longer be accepted.

**"Add cases from code" vs "Sync cases from code" were undistinguished, and one was described incorrectly.** They sat as adjacent bullets, the second claiming to import code-defined cases into the hosted dataset. Neither transfers anything. Both open a prefilled code snippet for you to run: `add_cases(...)` and `push_dataset(...)` respectively.

**The Datasets list was described as time-range-bounded for both kinds.** Hosted datasets are listed regardless of the selected time range. Only code-defined datasets are derived from experiments, so only those are bounded by it. The troubleshooting advice to widen the time range was therefore misleading for hosted datasets.

**The identity model was unstated.** The list is keyed on the dataset name, so a hosted dataset and a code-defined one sharing a name are a single row carrying both the hosted cases and the experiment history. This is the unstated reason "use stable names" keeps appearing as advice throughout these pages.

## What this adds

- A **"How cases get into a hosted dataset"** section: the three population paths (Live view, the Cases tab, the SDK) with their volume and when to reach for each, plus the 10,000-case ceiling.
- A **"Schemas are enforced on every write"** subsection: what is validated, the verbatim error format, that only present fields are checked (so a schema on `expected_output` or `metadata` does not make one mandatory), and that schema changes are not retroactive.
- Troubleshooting for: a case failing validation; the `Untitled` placeholder that collects every experiment reporting no dataset name; code-defined datasets disappearing once all their experiments are archived; and the list row cap, which makes search rather than scrolling the way to reach an old dataset.
- The identity/merge rule on the concept page.
- Repointed the SDK page's two "Web UI Guide" links at `manage-datasets.md`. They pointed at `datasets-and-experiments.md`, the concept page rather than the UI walkthrough.

## How the claims were verified

Every behavioral claim was checked by **executing the code**, not by reading it:

- **Frontend**, via vitest against the real modules: `getHostedAddCasesSnippet` emits `client.add_cases(...)` and `getHostedPushDatasetSnippet` emits `client.push_dataset(...)`, both prefilled with the dataset name and with type names taken from the schemas; `buildUnifiedDatasetsFromNames` collapses a hosted and a code-defined dataset of the same name into one `isHosted: true` row; `datasetDisplayName` renders the unnamed sentinel as `Untitled`.
- **Backend**, by running `_build_validators` / `_validate_case_against_schemas` under Python 3.14: a wrong-typed or missing required field is rejected with `Schema validation failed: inputs.question: 123 is not of type 'string'`; an absent or explicitly-null `expected_output` is accepted even when an output schema exists; a present-but-wrong one is rejected; with no schemas defined nothing is enforced. `MAX_CASES_PER_DATASET` is 10,000.
- The whole `src/app/evals` suite (77 files, 573 tests) passes on current `main`.

**Reviewer note, possible test gap:** no test in the backend suite asserts a schema-validation rejection. The raise sites carry `# pragma: full coverage`, and the case-limit path has a test (`'Dataset cannot exceed 10,000 cases'`) while the validation path does not. The behavior is real, but it is currently unguarded.

## Notes

- Prose only, no new Python examples, so the `test_docs` example checks are unaffected.
- These labels are what the `evals_workspace` flag produces. It is in `SHIPPED_FEATURE_FLAGS` and catalogued as a shipped fallback, so it defaults on; with it off the badge reads `Local` and the button reads `Push dataset from code`.
- I could not run a local stack in this environment: the egress policy returns 403 for both `us-docker.pkg.dev` and Docker Hub's blob CDN, so no images can be pulled or built. The verification above is what replaced it.
- I could not run the pre-commit hooks either: this environment has uv 0.8.17 and the repo requires >= 0.12.1. The hooks that apply to a Markdown-only change (`end-of-file-fixer`, `trailing-whitespace`) I verified by hand; the rest are Python- and workflow-scoped and match no file in this diff.
- One claim I deliberately did not make: what happens to an in-flight experiment when a schema is tightened mid-run, because I did not verify it.

Add the `trigger:docs` label if you want a rendered preview on this PR.

### CODE REVIEW

File: docs/evaluate/manage-datasets.md [hunk 1]
```diff
@@ -50,15 +59,48 @@ You can skip schemas and the first case. Add them later when the shape becomes c
...
-- **Add cases from code** shows the code needed to publish cases.
-- **Sync cases from code** imports code-defined cases into the hosted dataset.
+- **Add cases from code** opens a prefilled `add_cases(...)` snippet for appending or updating individual cases.
+- **Sync cases from code** opens a prefilled `push_dataset(...)` snippet for publishing a whole local dataset.
 - **Export** downloads the cases for reuse outside Logfire.
 - **Edit** changes the dataset name, description, or schemas.
 
 ![Edit the cases in a hosted dataset](../images/guide/evals/hosted-dataset-cases.webp)
 
+**Add cases from code** and **Sync cases from code** both open a code snippet for you to run. Neither transfers anything on its own. Both snippets arrive prefilled with the dataset's name and the type names taken from its schemas, so the difference between them is only which SDK call they hand you: `add_cases(...)` appends to what is already there, and `push_dataset(...)` publishes a local `Dataset` as a whole. See the [Datasets SDK](datasets-sdk.md) for both.
```
[the remark this answers is not in the document]
Comment (summerscope) [reply docs/evaluate/manage-datasets.md]: Confirmed against the SDK source and fixed in 6665164.

`push_dataset` is create-or-update on the dataset followed by `add_cases(..., on_conflict='update')` (`logfire/experimental/api_client.py:775`), so there is no delete path. I took the correction but worded it as two sentences rather than the suggestion, and stated the consequence explicitly: a case removed locally stays hosted, so the hosted set can drift from what the code defines. That drift seemed like the part a reader actually needs.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/evaluate/manage-datasets.md [hunk 2]
```diff
@@ -50,15 +59,48 @@ You can skip schemas and the first case. Add them later when the shape becomes c
...
+```text
+Schema validation failed: inputs.question: 123 is not of type 'string'
+```
+
+Two details matter when you plan a schema:
+
+- Only fields that are present are validated. A schema on `expected_output` or `metadata` does not force a case to carry one; it constrains the value when a case does.
+- Schemas are enforced from the moment you define them, but they are not applied retroactively. Cases that predate a schema stay as they are, and you find out they no longer match the next time something writes to them.
+
+Define schemas once the shape of a case has settled. Adding them early to a dataset you are still exploring turns every subsequent write into a validation error.
```
[the remark this answers is not in the document]
Comment (summerscope) [reply docs/evaluate/manage-datasets.md]: Fair catch, fixed in 6665164. "Turns every subsequent write into a validation error" was an overstatement: only writes whose present values violate the schema are rejected, which I had already verified by running `_validate_case_against_schemas` directly.

Reworded rather than taking the suggestion verbatim, to keep the actual advice (add cases first, describe them once the pattern is clear) instead of only restating the mechanics.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/evaluate/manage-datasets.md [hunk 3]
```diff
@@ -50,15 +59,42 @@ You can skip schemas and the first case. Add them later when the shape becomes c
...
 If Logfire already discovered a code-defined dataset with the same name, creating its hosted counterpart can import the latest cases instead of starting empty. Review the imported cases before relying on them as a shared test set.
 
+## How cases get into a hosted dataset
+
+Cases reach a hosted dataset three ways. They combine freely in one dataset:
+
+| Path | Volume | Use it when |
+| --- | --- | --- |
+| [Live view](#add-a-case-from-a-production-trace) | One case per span | A real request is worth keeping as a regression case |
+| [The Cases tab](#create-a-hosted-dataset) | One case at a time | You are hand-writing a specific edge case |
```
[the remark this answers is not in the document]
Comment (summerscope) [reply docs/evaluate/manage-datasets.md]: Right, that row contradicted the paragraph above it. Fixed in 34fb223.

Scoped the row to the manual **Add case** button rather than widening the volume column, so the table stays three distinct paths and the SDK row keeps sole ownership of "Bulk". Added a line noting the Cases tab reaches the SDK path through those snippets, so bulk work runs from your machine rather than in the browser.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/evaluate/manage-datasets.md [hunk 4]
```diff
@@ -50,15 +59,50 @@ You can skip schemas and the first case. Add them later when the shape becomes c
...
+- **Sync cases from code** opens a prefilled `push_dataset(...)` snippet for publishing a whole local dataset.
 - **Export** downloads the cases for reuse outside Logfire.
 - **Edit** changes the dataset name, description, or schemas.
 
 ![Edit the cases in a hosted dataset](../images/guide/evals/hosted-dataset-cases.webp)
 
+**Add cases from code** and **Sync cases from code** both open a code snippet for you to run. Neither transfers anything on its own. Both arrive prefilled with the dataset's name and the type names taken from its schemas, so the difference is only which SDK call they hand you. `add_cases(...)` adds the cases you pass, updating any that match an existing case name. `push_dataset(...)` pushes a whole local `Dataset`: it creates or updates the hosted dataset's schemas and evaluators, then upserts the cases you passed.
+
+Neither call deletes anything. A case you remove from your local dataset stays in the hosted one until you delete it there, so a hosted dataset can accumulate cases your code no longer defines. See the [Datasets SDK](datasets-sdk.md) for both.
```
[the remark this answers is not in the document]
Comment (summerscope) [reply docs/evaluate/manage-datasets.md]: Correct, and my own fault: the previous commit fixed one overstatement by introducing another. Fixed in ed5423a.

Confirmed against `push_dataset`'s docstring (`api_client.py:783`) and the round-tripping note already in the SDK guide: case-level evaluators go up with their cases, dataset-level `evaluators` and `report_evaluators` go up with the dataset, and each push overwrites the hosted values.

Scoped the no-delete statement to cases and gave the evaluator exception its own paragraph, rather than folding both into one sentence, since the two have opposite consequences and a reader skimming for "what can I lose" should not have to parse a contrast clause to find it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/evaluate/manage-datasets.md [hunk 5]
```diff
@@ -50,15 +59,48 @@ You can skip schemas and the first case. Add them later when the shape becomes c
...
+
+| Path | Volume | Use it when |
+| --- | --- | --- |
+| [Live view](#add-a-case-from-a-production-trace) | One case per span | A real request is worth keeping as a regression case |
+| [The Cases tab](#create-a-hosted-dataset) | One case at a time | You are hand-writing a specific edge case |
+| [The SDK](datasets-sdk.md) | Bulk | Cases are generated, migrated, or already in code |
+
+Adding from Live view is the path that turns production behavior into test cases, and it is usually where a shared dataset starts. The SDK is the only path that scales to many cases at once.
+
+A hosted dataset holds at most **10,000 cases**. Adding a case beyond that limit is rejected, including in bulk, so a large import needs splitting across datasets.
```
[the remark this answers is not in the document]
Comment (summerscope) [reply docs/evaluate/manage-datasets.md]: Valid, and fixed across both files in c9b66ec.

Checked the upsert path before changing anything: it computes `new_case_count` as the named cases that do not already exist plus every unnamed case, then compares that against the ceiling. So updates to existing named cases genuinely do not consume capacity, and a dataset sitting at 10,000 can still be re-pushed. The old wording would have told someone with a full dataset they could never push again.

Both guides now say the ceiling counts the cases a write would create, and both note the corollary you didn't mention: an unnamed case always counts as new, which is why repeated pushes of unnamed cases grow a dataset.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/evaluate/manage-datasets.md [hunk 6]
```diff
@@ -7,7 +7,16 @@ description: "Create a hosted or code-defined evaluation dataset, curate its cas
...
 Build a dataset, a stable collection of test cases that you can run repeatedly as your agent, model, or prompt changes.
 
-Open a project, then select **AI Evaluations** > **Datasets & experiments**. The **Datasets** tab shows hosted and code-defined datasets in the selected time range.
+Open a project, then select **AI Evaluations** > **Datasets & experiments**. The **Datasets** tab lists both kinds of dataset together.
+
+## Understand what the list contains
+
+The list combines two sources, keyed on the dataset name:
+
+- **Hosted** datasets are records stored on Logfire. They are always listed, whatever time range you have selected.
```
[the remark this answers is not in the document]
Comment (summerscope) [reply docs/evaluate/manage-datasets.md:16]: Half right, so I fixed it at the other end. Addressed in c9b66ec.

"Always listed" is accurate as written: `list_datasets` in the DAO selects hosted datasets for the project with `ORDER BY d.name ASC` and **no** `LIMIT`. The cap is `DATASET_LIST_LIMIT`, which is passed only to the experiment-derived name query, so it bounds the code-defined half of the list and never the hosted half.

The genuine ambiguity was the troubleshooting entry you're pointing at: it described the cap without saying which half it governs, so a reader could carry it back to line 16. I scoped that entry to code-defined datasets and stated that hosted ones are not capped, rather than weakening a statement that is true.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/evaluate/manage-datasets.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (summerscope) [reply docs/evaluate/manage-datasets.md]: Agreed, and already fixed in 6665164, which landed before this review was posted (it reviewed `f68b475`). The line now reads: "`add_cases(...)` adds the cases you pass, updating any that match an existing case name."

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/evaluate/manage-datasets.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (summerscope) [reply docs/evaluate/manage-datasets.md]: Agreed, and already fixed in 6665164, which landed before this review was posted (it reviewed `f68b475`). Macroscope raised the same point independently. The line now reads: "While the shape is still moving, a schema rejects each write that has drifted from it, so it is usually easier to add cases first and describe them once the pattern is clear."

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by summerscope** [comment]
@coderabbitai review

All four findings are addressed and the branch has stopped moving, so this should be the final head (`c9b66ec`). Two of them (`add_cases` conflict behavior, the schema warning) were already fixed in `6665164` before your review of `f68b475` posted; the 10,000-case qualification and the list-cap scoping landed in `c9b66ec`.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by summerscope** [comment]
## Explicitly accepting the Moderate merge-risk note

The final CodeRabbit review of `c9b66ec` reports **"No actionable comments were generated"** and 4/4 pre-merge checks, but its aggregate risk line says the docs "still contain inaccurate guidance about case updates, dataset limits, schema validation, and dataset listing behavior."

Those are the same four findings raised earlier in this PR. Each was either fixed and then confirmed by CodeRabbit, or withdrawn by it. The risk line summarizes the findings history rather than the current text, so I have not changed the docs again: doing so would make them wrong.

Current state of each, with the evidence:

| Topic | Current text | Basis |
| --- | --- | --- |
| Case updates | "`add_cases(...)` adds the cases you pass, updating any that match an existing case name" | `on_conflict='update'` default; CodeRabbit marked this thread confirmed |
| Dataset limits | ceiling counts "the cases a write would create"; updates to existing names do not consume capacity; an unnamed case always counts as new | `new_case_count = len(named_cases) - existing_name_count + unnamed_case_count`; CodeRabbit marked this thread confirmed |
| Schema validation | enforced on every write; only fields that are **present** are validated; not retroactive | ran `_build_validators` / `_validate_case_against_schemas` directly: absent and null `expected_output` accepted, present-but-wrong rejected; error string quoted verbatim from that run |
| Dataset listing | hosted datasets always listed; the row cap applies to the code-defined half only | `list_datasets` selects with `ORDER BY d.name ASC` and no `LIMIT`; `DATASET_LIST_LIMIT` is passed only to the experiment-derived query. CodeRabbit **withdrew** this finding and recorded a learning agreeing |

Every behavioral claim in this PR was checked by executing the code rather than reading it. The one thing I could not do is exercise the running product: this environment's egress policy blocks the container registries, so no local stack was possible. That limitation is in the PR description.

Flagging for the human reviewer rather than resolving silently, since the risk line is what shows on the PR.

---
_Generated by [Claude Code](https://claude.ai/code)_
